improve documentation (#110)
legend-exp/legend-pygeom-optics: docs/source/user_guide.md

- TPB: {func}`legendoptics.tpb.pyg4_tpb_attach_rindex`,
  {func}`legendoptics.tpb.pyg4_tpb_attach_wls`
- Fibers: {func}`legendoptics.fibers.pyg4_fiber_core_attach_rindex`,
  {func}`legendoptics.fibers.pyg4_fiber_core_attach_absorption`,
  {func}`legendoptics.fibers.pyg4_fiber_core_attach_wls`,
  {func}`legendoptics.fibers.pyg4_fiber_core_attach_scintillation` (plus
  cladding 1/2 rindex attachers)
- Reflectors: {func}`legendoptics.tetratex.pyg4_tetratex_attach_reflectivity`,
  {func}`legendoptics.tyvek.pyg4_tyvek_attach_reflectivity`,
  {func}`legendoptics.vm2000.pyg4_vm2000_attach_wls` (see module for more:
  rindex, reflectivity, border params)
- Substrates/metals:
  {func}`legendoptics.copper.pyg4_copper_attach_reflectivity`,
  {func}`legendoptics.germanium.pyg4_germanium_attach_reflectivity`
- Semiconductors/glass:
  {func}`legendoptics.silicon.pyg4_silicon_attach_complex_rindex`,
  {func}`legendoptics.silica.pyg4_silica_attach_rindex`
- Others: {mod}`legendoptics.nylon`, {mod}`legendoptics.ultem`,
  {mod}`legendoptics.water`, {mod}`legendoptics.pmts`

Minimal usage outline (pseudo-code):

```python
# Minimal outline with pyg4ometry (exact construction details may vary):

> import pyg4ometry.geant4 as g4

Reviewer: ```suggestion
import pint
import pyg4ometry.geant4 as g4
```

from legendoptics.lar import (
    pyg4_lar_attach_rindex,
    pyg4_lar_attach_attenuation,
    pyg4_lar_attach_scintillation,
)

> reg = g4.Registry()

Reviewer: ```suggestion
u = pint.get_application_registry().get()
reg = g4.Registry()
```

# Create your material in pyg4ometry here (example only; adjust to your setup)
lar_mat = g4.Material(name="LAr", density=1.396, registry=reg)  # density in g/cm**3

# Attach optical properties
pyg4_lar_attach_rindex(lar_mat, reg)

> pyg4_lar_attach_attenuation(lar_mat, reg, lar_temperature=88.8)

Reviewer: ```suggestion
pyg4_lar_attach_attenuation(lar_mat, reg, lar_temperature=88.8 * u.K)
```

pyg4_lar_attach_scintillation(lar_mat, reg)
```

Under the hood, the helpers use {mod}`legendoptics.pyg4utils` to patch
{mod}`pyg4ometry` with pint-aware methods:

- `Material.addVecPropertyPint(name, energy, values)`
- `Material.addConstPropertyPint(name, value)`

These ensure proper unit handling and ascending-energy ordering.

## Modifying properties without forking (pluggable store)

Many property functions are decorated with
{func}`legendoptics.store.register_pluggable`. This makes them dynamically
replaceable at runtime.

Common operations:

```python
from legendoptics import store
from legendoptics.pen import pen_refractive_index

# Replace a property implementation
def my_pen_rindex() -> float:
    return 1.52  # new constant refractive index

pen_refractive_index.replace_implementation(my_pen_rindex)

# Query which functions were replaced
assert "pen_refractive_index" in store.get_replaced()

# Reset just this function
pen_refractive_index.reset_implementation()

# Or reset everything back to defaults
store.reset_all_to_original()
```

> Note: Any function listed in the module’s `__all__` or imported via
> `legendoptics.<submodule>` that is decorated with `@store.register_pluggable`

Reviewer: ```suggestion
Note: Any imported function that is decorated with `@store.register_pluggable`
```

def mymat_refractive_index() -> float:
    return 1.37

@store.register_pluggable
def mymat_absorption() -> tuple[u.Quantity, u.Quantity]:
    λ, L = readdatafile("mymat_absorption.dat")  # "# unit1 unit2" header expected
    return λ, L
```

2. Implement pyg4 attachers

- Sample wavelengths with {func}`legendoptics.pyg4utils.pyg4_sample_λ` where
  needed.
- Convert wavelength to energy inside a pint context and attach via
  `addVecPropertyPint` and `addConstPropertyPint`.

Example:

```python
import numpy as np
from legendoptics.pyg4utils import pyg4_sample_λ
import pint

u = pint.get_application_registry()

def pyg4_mymat_attach_rindex(mat, reg):
    λ = np.array([650.0, 115.0]) * u.nm
    r = [mymat_refractive_index()] * 2
    with u.context("sp"):
        mat.addVecPropertyPint("RINDEX", λ.to("eV"), r)

def pyg4_mymat_attach_absorption(mat, reg):
    λ, L = mymat_absorption()
    with u.context("sp"):
        mat.addVecPropertyPint("ABSLENGTH", λ.to("eV"), L)
```

3. Optional: WLS and scintillation

- Follow patterns in `pen.py`, `tpb.py`, `fibers.py`, and `lar.py`:
  - WLS: `WLSABSLENGTH`, `WLSCOMPONENT`, `WLSTIMECONSTANT`, optional
    `WLSMEANNUMBERPHOTONS`

>   - Scintillation: define a {class}`legendoptics.scintillate.ScintConfig` in
>     `scintillate.py` style and use
>     {func}`legendoptics.pyg4utils.pyg4_def_scint_by_particle_type`.

Reviewer: ```suggestion
  - Scintillation: define a {class}`legendoptics.scintillate.ScintConfig` and
    use {func}`legendoptics.pyg4utils.pyg4_def_scint_by_particle_type`.
```

4. Include data files

> - Place spectral data in an importable package (e.g., `legendoptics.data`
>   style).

Reviewer: ```suggestion
- Place spectral data in an importable package (e.g., the `data` directory in
  the pakage `mypkg` is importable as `mypkg.data`)
```

>   the pakage `mypkg` is importable as `mypkg.data`)

Reviewer: ```suggestion
  the package `mypkg` is importable as `mypkg.data`)
```

- Format: first line header with units (`# unit1 unit2`), then pairs of numbers;
  comments allowed after `#` (after the header line).

## CLI helper

A small CLI (defined in {mod}`legendoptics.cli`) can generate
G4GeneralParticleSource emission spectra:

```
legend-pygeom-optics g4gps lar_emission out.mac
legend-pygeom-optics g4gps pen_emission out.mac
legend-pygeom-optics g4gps fiber_emission out.mac
```

This uses the same emission spectra as the attachers.

## Practical examples

> - pygeomtools (central material defintions):

Reviewer: ```suggestion
- pygeomtools (central material definitions):
```

  needed.
- Convert wavelength to energy inside a pint context and attach via
  `addVecPropertyPint` and `addConstPropertyPint`.

Example:

```python
import numpy as np
from legendoptics.pyg4utils import pyg4_sample_λ
import pint

u = pint.get_application_registry()

def pyg4_mymat_attach_rindex(mat, reg):
    λ = np.array([650.0, 115.0]) * u.nm
    r = [mymat_refractive_index()] * 2
    with u.context("sp"):
        mat.addVecPropertyPint("RINDEX", λ.to("eV"), r)

def pyg4_mymat_attach_absorption(mat, reg):
    λ, L = mymat_absorption()
    with u.context("sp"):
        mat.addVecPropertyPint("ABSLENGTH", λ.to("eV"), L)
```

3. Optional: WLS and scintillation

- Follow patterns in `pen.py`, `tpb.py`, `fibers.py`, and `lar.py`:
  - WLS: `WLSABSLENGTH`, `WLSCOMPONENT`, `WLSTIMECONSTANT`, optional
    `WLSMEANNUMBERPHOTONS`
  - Scintillation: define a {class}`legendoptics.scintillate.ScintConfig` in
    `scintillate.py` style and use
    {func}`legendoptics.pyg4utils.pyg4_def_scint_by_particle_type`.

4. Include data files

- Place spectral data in an importable package (e.g., `legendoptics.data`
  style).

> - Format: first line header with units, then pairs of numbers; comments allowed
>   after `#`.

Reviewer: ```suggestion
- Format: first line header with units (`# unit1 unit2`), then pairs of numbers;
  comments allowed after `#` (after the header line).
```

## CLI helper

A small CLI (defined in {mod}`legendoptics.cli`) can generate
G4GeneralParticleSource emission spectra:

```
legend-pygeom-optics g4gps lar_emission out.mac
legend-pygeom-optics g4gps pen_emission out.mac
legend-pygeom-optics g4gps fiber_emission out.mac
```

This uses the same emission spectra as the attachers.

## Practical examples

> - LEGEND-200:
>   [legend-exp/legend-pygeom-l200](https://github.com/legend-exp/legend-pygeom-l200)
> - LEGEND-1000:
>   [legend-exp/legend-pygeom-l1000](https://github.com/legend-exp/legend-pygeom-l1000)
> 
> Both repositories show how materials are defined, how attachers are called
> during geometry construction, and how to manage optical surfaces and properties
> consistently across a large detector model.

Reviewer: ```suggestion
- pygeomtools (central material defintions):
  [legend-exp/legend-pygeom-tools](https://github.com/legend-exp/legend-pygeom-tools)
- LEGEND-200:
  [legend-exp/legend-pygeom-l200](https://github.com/legend-exp/legend-pygeom-l200)
- LEGEND-1000:
  [legend-exp/legend-pygeom-l1000](https://github.com/legend-exp/legend-pygeom-l1000)

The three repositories illustrate how materials are defined with their attachers,
and how to manage optical surfaces and properties consistently across a large
detector model.
```
